### PR TITLE
fix(ActiveShape): Fix breaking change in ActiveShape type 

### PR DESCRIPTION
**Pull request by andrewangelle**
Fix ActiveShape type that causes  breaking change in Bar shape prop

## Description

Extend the definition of `ActiveShape` to allow consumer to define the type of props argument on their own render prop being passed to `shape`.

## Related Issue
#3911 

## How Has This Been Tested?
unit tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/chart/BarChart.spec.tsx [hunk 1]
```diff
@@ -265,7 +271,7 @@ describe('<BarChart />', () => {
   });
 
   test('Render customized shapem when shape is set to be a function', () => {
-    const renderShape = (props: any) => {
+    const renderShape = (props: RectangleProps & DataType) => {
```
Comment (andrewangelle) [inline test/chart/BarChart.spec.tsx:274]: This is very similar to the implementation provided by the user in the reproduction linked in the issue. 

This also would throw a type error prior to the change in this PR.

File: src/util/types.ts [hunk 2]
```diff
@@ -1320,5 +1321,6 @@ export type Size = { width: number; height: number };
 export type ActiveShape<PropsType = Record<string, any>, ElementType = SVGElement> =
   | ReactElement<SVGProps<ElementType>>
   | ((props: PropsType) => ReactElement<SVGProps<ElementType>>)
+  | ((props: unknown) => JSX.Element)
```
Comment (andrewangelle) [inline src/util/types.ts:1324]: This is the fix

**Comment by ckifer** [comment]
So now if someone has `activeShape={props => <Trapezoid {...props} />}` they are most likely going to get the implicit `any` errors that just appeared in our CI no?

**Review by ckifer — APPROVED** [review]
Definitely better either way and fixes the reported bug, might cause some other tsc errors for folks, but those should be easy for them to fix. Merging this for now

**Comment by andrewangelle** [comment]
> So now if someone has `activeShape={props => <Trapezoid {...props} />}` without an explicit type they are most likely going to get the implicit `any` errors that just appeared in our CI no?

Yes. But that was the behavior prior to `2.9.0`, so we're likely fixing more folks than we're breaking. For the record, I'd like to remove this in the next major version. Assuming our types are accurate, I think we should be throwing a type error if the user is assigning a random type to a parameter that we internally define.

**Comment by ckifer** [comment]
Agreed but we should take a closer look to ensure the type doesn't leave anything out before that

**Comment by ckifer** [comment]
Hmm, If I cherry pick this commit from `v2.9.0` tag I get a bunch of compilation failures
```
No overload matches this call.
  Overload 1 of 2, '(props: Props | Readonly<Props>): Bar', gave the following error.
    Type '(props: RectangleProps & DataType) => React.JSX.Element' is not assignable to type 'ActiveShape<BarProps, SVGPathElement> | undefined'.
      Type '(props: RectangleProps & DataType) => React.JSX.Element' is not assignable to type '(props: unknown) => Element'.
        Types of parameters 'props' and 'props' are incompatible.
          Type 'unknown' is not assignable to type 'Omit<SVGProps<SVGPathElement>, "radius"> & RectangleProps & DataType'.
            Type 'unknown' is not assignable to type 'Omit<SVGProps<SVGPathElement>, "radius">'.
  Overload 2 of 2, '(props: Props, context: any): Bar', gave the following error.
    Type '(props: RectangleProps & DataType) => React.JSX.Element' is not assignable to type 'ActiveShape<BarProps, SVGPathElement> | undefined'.
```

Any idea?

**Comment by ckifer** [comment]
@andrewangelle https://github.com/ckifer/recharts/blob/patch/2.9.1/test/chart/BarChart.spec.tsx#L294

I pushed my patch branch up.. I can't figure out why that type isn't working
The difference I see is 

`(property) shape?: ActiveShape<BarProps, SVGPathElement>` (working) vs `(property) shape?: ActiveShape<BarProps, SVGPathElement> | undefined` (not working), but no idea why

**Comment by andrewangelle** [comment]
@ckifer I can't track down why master has no errors, but cherry-picked into 2.9.0 does.

It does seem like `| ((props: any) => JSX.Element)` instead of `| ((props: unknown) => JSX.Element)` removes the type errors. Do you want me to PR that?

**Comment by ckifer** [comment]
Hmm, no worries I can just change it and publish that for now. I don't have plans to merge 2.9.1 back into master or do another patch release if I can help it. There's been too much other stuff merged since. Thank you!

**Pull request merged**